chore!: Remove open keyword from Noir

spalladino opened the pull request:
Removes the unused `open` keyword from Noir, as requested [here](https://github.com/AztecProtocol/aztec-packages/pull/4898#issuecomment-1978566866).

**Breaking change**: This PR changes the format of Noir contract build artifacts, so a full rebuild of contracts is needed after this PR is merged. No other action is needed by the user. If they fail to do it, they'll see a message along the lines of _No custom attributes found for contract function foo. Try rebuilding the contract with the latest nargo version._ (cc @AztecProtocol/devrel @rahul-kothari)

spalladino committed:
chore: Remove open keyword from Noir

TomAFrench commented:
Is open completely unused by aztec then? If so can you remove it from the compiler?

spalladino commented:
I understand the `is_open` flag for a function is still in use, as it affects its target (ACIR vs Brillig/AVM) (@Maddiaa0 please correct me if I'm wrong). The only difference is that this is now specified by using the `aztec(public)` macro over the `open` keyword.

TomAFrench committed:
chore: remove knowledge of aztec function types from compiler (#4975)

This PR removes the concept of a function being internal or "open" from
the Noir compiler as these are now concepts which are aztec specific. We
instead attach all of a contract functions custom attributes to it in
the build artifact so aztec can parse these to determine whether a
function is open or secret, etc.

@spalladino This follows on from #4967 but I've made this as a separate
PR into yours as I don't want to hijack it.

spalladino committed:
Clean up more internal stuff

spalladino committed:
Fix AVM build

spalladino committed:
Add is_unconstrained flag to function output

spalladino committed:
Update function to load Noir artifact

spalladino commented:
@TomAFrench I've had to add an `is_unconstrained` flag to the function output. The `custom_attributes` alone were not enough to differentiate between an "unconstrained contract function" (ie a getter, which is signalled by marking the function with the `unconstrained` noir keyword) and a manually-written private function (ie one written manually and without the usage of macros). Let me know what you think.

Long term, maybe a better approach would be to require all contract functions to be decorated with one of `aztec(private|public|getter)`, and functions that do not have any of these modifiers are deemed internal to the contract (what today is `contract_library_function`).

spalladino committed:
Add new is_unconstrained flag to transpiler

spalladino committed:
Lint

spalladino commented:
@TomAFrench this PR is failing because some contracts started exceeding the max bytecode length. I checked and indeed public functions are now about 10-15% bigger. For instance, in the token contract compiled in this PR:

```
$ yarn aztec-cli inspect-contract TokenContractArtifact | grep -A3 transfer
open transfer_public(from: (inner: Field), to: (inner: Field), amount: Field, nonce: Field) 
	function signature: transfer_public((Field),(Field),Field,Field)
	selector: 0x7db8f449
	bytecode: 22515 bytes (sha256 81af6e02824d206d4a1d87e084bb3ff6bf04bfabae2686ac0d5379c00f7250fa)
--
secret transfer(from: (inner: Field), to: (inner: Field), amount: Field, nonce: Field) 
	function signature: transfer((Field),(Field),Field,Field)
	selector: 0xb1aa668a
	bytecode: 694454 bytes (sha256 8a01f0f8da7f8aba6551de8c2bd4f42e1176eca140a39bb631d42f6ac7bd5698)
```

Whereas in master:

```
$  yarn aztec-cli inspect-contract TokenContractArtifact | grep -A3 transfer
(node:73168) ExperimentalWarning: Importing JSON modules is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
secret transfer(from: (inner: Field), to: (inner: Field), amount: Field, nonce: Field) 
	function signature: transfer((Field),(Field),Field,Field)
	selector: 0xb1aa668a
	bytecode: 693619 bytes (sha256 bb9b31696b92afdfb36e8aff8f7704689fcdce932f177c1cabad08931363fda1)
--
open transfer_public(from: (inner: Field), to: (inner: Field), amount: Field, nonce: Field) 
	function signature: transfer_public((Field),(Field),Field,Field)
	selector: 0x7db8f449
	bytecode: 19712 bytes (sha256 e520866d0d4c5655b48e4aa6a79f3182aaa74514e20a99de238c1f019916df6b)
```

Is it possible that by removing the `open` flag in functions we're accidentally compiling public functions as ACIR instead of Brillig? How could I verify this easily? If not, what do you think could be the issue?

TomAFrench commented on yarn-project/types/src/abi/contract_artifact.ts:
```diff
@@ -125,6 +135,19 @@ function generateFunctionArtifact(fn: NoirCompiledContractFunction): FunctionArt
+    return FunctionType.SECRET;
+  } else if (fn.custom_attributes.includes(AZTEC_PUBLIC_ATTRIBUTE)) {
+    return FunctionType.OPEN;
+  } else if (fn.is_unconstrained) {
+    return FunctionType.UNCONSTRAINED;
+  } else {
+    // Default to a private function (see simple_macro_example_expanded for an example of this behavior)
+    return FunctionType.SECRET;
+  }
```
One janky solution to this is just check to see if the function is a single opcode. Dealing with the private context ensures that no secret function can be a single opcode so this should be a sufficient check.

TomAFrench commented:
@spalladino, the selectors of the two functions are swapping between the two branches, is that expected? I'm wondering if there's some confounding factor here as I can't see anything that would result in an `open` function being compiled to acir.

TomAFrench commented:
I've compiled the contract

```
contract Counter {
    use dep::aztec::prelude::{AztecAddress, Map};
    use dep::aztec::context::Context;
    use dep::value_note::{balance_utils, value_note::{ValueNote, VALUE_NOTE_LEN}};
    use dep::easy_private_state::EasyPrivateUint;

    #[aztec(open)]
    fn get_counter(owner: AztecAddress) -> pub Field {
        assert(owner == AztecAddress::empty());
        2
    }
}
```

This results in the output
```
Compiled ACIR for Counter::get_counter (unoptimized):
current witness index : 1
public parameters indices : []
return value indices : [1]
EXPR [ (1, _0) 0 ]
EXPR [ (-1, _1) 2 ]
```

So it does seem like we're compiling to ACIR for some reason rather than brillig.

TomAFrench commented:
I've compiled the contract 

```rust
contract Counter {
    // docs:start:imports
    use dep::aztec::prelude::{AztecAddress, Map};
    use dep::aztec::context::Context;
    use dep::value_note::{balance_utils, value_note::{ValueNote, VALUE_NOTE_LEN}};
    use dep::easy_private_state::EasyPrivateUint;
    // docs:end:imports

    #[aztec(private)]
    #[aztec(initializer)]
    fn constructor(headstart: u64, owner: AztecAddress) {
        // let counters = storage.counters;
        // counters.at(owner).add(headstart, owner);
    }

   
    #[aztec(public)]
    fn get_counter(owner: AztecAddress) -> Field {
        assert(owner == AztecAddress::empty());
        2
    }
 }
```

This gives the snipped output below which consists of a single brillig opcode so we're definitely not accidentally compiling it as if it were constrained.

```
Compiled ACIR for Counter::get_counter (unoptimized):
current witness index : 238
public parameters indices : []
return value indices : [40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 222, 223, 224, 225, 226, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 237, 238]
BRILLIG: inputs: [Single(Expression { mul_terms: [], linear_combinations: [(1, Witness(0))], q_c: 0 }), Single(Expression { mul_terms: [], linear_combinations: [(1, Witness(1))], q_c: 0 }), Single(Expression { mul_terms: [], linear_combinations: [(1, Witness(2))], q_c: 0 }), Single(Expression { mul_terms: [], linear_combinations: [(1, Witness(3))], q_c: 0 }), Single(Expression { mul_terms: [], linear_combinations: [(1, Witness(4))], q_c: 0 }), Single(Expression { mul_terms: [], linear_combinations: [(1, Witness(5))], q_c: 0 }), Single(Expression { mul_terms: [], linear_combinations: [(1, Witness(6))], q_c: 0 }), Single(Expression { mul_terms: [], linear_combinations: [(1, Witness(7))], q_c: 0 }), Single(Expression { mul_terms: [], linear_combinations: [(1, Witness(8))], q_c: 0 }), Single(Expression { mul_terms: [], linear_combinations: [(1, Witness(9))], q_c: 0 }), Single(Expression { mul_terms: [], linear_combinations: [(1, Witness(10))], q_c: 0 }), Array([Expression { mul_terms: [], linear_combinations: [(1, Witness(11))], q_c: 0 }, Expression { mul_terms: [], linear_combinations: [(1, Witness(12))], q_c: 0 }]), Array([Expression { mul_terms: [], linear_combinations: [(1, Witness(13))], q_c: 0 }, Expression { mul_terms: [], linear_combinations: [(1, Witness(14))], q_c: 0 }]), Array([Expression { mul_terms: [], linear_combinations: [(1, Witness(15))], q_c: 0 }, Expression { mul_terms: [], linear_combinations: [(1, Witness(16))], q_c: 0 }]), Single(Expression { mul_terms: [], linear_combinations: [(1, Witness(17))], q_c: 0 }), Single(Expression { mul_terms: [], linear_combinations: [(1, Witness(18))], q_c: 0 }), Single(Ex.....
```

TomAFrench commented:
What's counted as "contract bytecode"? We're obviously putting more data into the JSON file artifact but I'm assuming you're not just dumping that directly onchain right?

spalladino commented:
> What's counted as "contract bytecode"? We're obviously putting more data into the JSON file artifact but I'm assuming you're not just dumping that directly onchain right?

No, contract bytecode is the concatenation of the selector, internal flag, bytecode length, and bytecode for each public function:
```
export function packBytecode(publicFns: ContractClass['publicFunctions']): Buffer {
  return serializeArrayOfBufferableToVector(
    publicFns.map(fn => serializeToBuffer(fn.selector, fn.isInternal, numToInt32BE(fn.bytecode.length), fn.bytecode)),
  );
}
```

From `circuits.js/src/contract/public_bytecode.ts`

TomAFrench commented on yarn-project/types/src/abi/contract_artifact.ts:
```diff
@@ -125,6 +135,19 @@ function generateFunctionArtifact(fn: NoirCompiledContractFunction): FunctionArt
+function getFunctionType(fn: NoirCompiledContractFunction): FunctionType {
+  if (fn.custom_attributes.includes(AZTEC_PRIVATE_ATTRIBUTE)) {
+    return FunctionType.SECRET;
+  } else if (fn.custom_attributes.includes(AZTEC_PUBLIC_ATTRIBUTE)) {
```
```suggestion
  } else if (fn.custom_attributes.includes(AZTEC_PUBLIC_VM_ATTRIBUTE)) {
```

Here's the issue. Previously the `#[aztec(public-vm)]` attribute would set the `is_open` flag but now we're looking for the `#[aztec(public)]` attribute when selecting functions to run on the AVM.

TomAFrench commented on yarn-project/types/src/noir/index.ts:
```diff
@@ -7,8 +7,10 @@ import {
-/** The Aztec.nr function types. */
-type NoirFunctionType = 'Open' | 'Secret' | 'Unconstrained';
+export const AZTEC_PRIVATE_ATTRIBUTE = 'aztec(private)';
+export const AZTEC_PUBLIC_ATTRIBUTE = 'aztec(public)';
```
```suggestion
export const AZTEC_PUBLIC_ATTRIBUTE = 'aztec(public)';
```
```suggestion
export const AZTEC_PUBLIC_VM_ATTRIBUTE = 'aztec(public-vm)';
```

TomAFrench commented on yarn-project/types/src/abi/contract_artifact.ts:
```diff
@@ -7,7 +7,12 @@ import {
+import {
+  AZTEC_INTERNAL_ATTRIBUTE,
+  AZTEC_PRIVATE_ATTRIBUTE,
+  AZTEC_PUBLIC_ATTRIBUTE,
```
```suggestion
  AZTEC_PUBLIC_VM_ATTRIBUTE,
```

spalladino commented:
Sorry, I'm a bit confused: isn't the `public-vm` attribute a new one, used for AVM transpilation, and still not used in the noir-contracts?

TomAFrench commented:
I'm confused. I thought that open functions were functions that ran in the open context (i.e. on the public VM) what do they actually do?

TomAFrench commented on avm-transpiler/src/transpile_contract.rs:
```diff
@@ -82,7 +81,9 @@ impl From<CompiledAcirContract> for TranspiledContract {
         for function in contract.functions {
             // TODO(4269): once functions are tagged for transpilation to AVM, check tag
-            if function.function_type == ContractFunctionType::Open
+            if function
```
This implies that open is public vm.

spalladino commented:
> I'm confused. I thought that open functions were functions that ran in an open context (i.e. on the public VM) what do they actually do?

The public VM doesn't exist yet, so today public functions are just Brillig that gets run on the sequencer, but not proven.

spalladino commented:
Got it, I'll look it from Aztec's side on Monday. Thanks for looking into this Tom!

spalladino committed:
Fix introducing storage definition twice in public functions

spalladino commented:
@TomAFrench found it: https://github.com/AztecProtocol/aztec-packages/pull/4967/commits/4550854bf39d09aaeeb333cd7094688fe289ac21

We were introducing the storage definition into public functions twice. We should be good to merge now if tests pass.

TomAFrench commented on yarn-project/types/src/noir/index.ts:
```diff
@@ -40,10 +42,10 @@ export interface NoirFunctionAbi {
-  is_internal: boolean;
+  /** Whether the function is unconstrained. */
+  is_unconstrained: boolean;
+  /** Attributes injected via macros. */
```
```suggestion
  /** Custom attributes attached to function */
```

These aren't injected by the macros

TomAFrench committed:
Update yarn-project/types/src/noir/index.ts

TomAFrench commented:
Looks like circuits.js isn't recompiling some contracts it's using?

spalladino committed:
Load contract artifact for testing from noir-contracts instead of committing

spalladino committed:
Fix snapshot

spalladino committed:
Check public-vm attribute when processing new ABI format
